Merge up to 88b9bd396d5a1f83c49fa1b28745fd7feaca2b2e from  upstream

en-sc opened the pull request:
Conflicts:

* `configure.ac`: commit bd5649dfbe10a79f72bee2ee48cf24c10de7b5d8 ("Support cJTAG JScan3 mode") and commit 7d0e125896291e7683f76aa1ad10532bb192692b ("jtag/drivers: Add support for CH347-based JTAG adapters")
* `src/jtag/drivers/ftdi.c`: commit f61098c2258c2ca81fd7c0ed943c36f6a82c5d49 ("Rename ftdi_oscan1 to ftdi_cjtag") and commit 867611d3e19502d3affa780a38de3b41201ee712 ("jtag: drivers: align switch and case statements") -- resolved by aligning the statements.
* `src/rtos/freertos.c`: commit 50a5971be27ff11061776afc8014bbff22870b7c ("RISC-V Freertos support (#582)") and commit commit 3954896d6e41298efd74e8bb7af60ec49efae809 ("rtos/FreeRTOS: fix next pointer member offset in FreeRTOS lists") -- dropped the upstream change since the offsets are dynamically computed by `freertos_compute_offsets()`.
* `src/rtos/freertos.c`:  due to commit b3b790e4e01331f79b196f8312193ae0130d2394 ("rtos: rework rtos_create()") -- the change is already adopted.
* `src/rtos/hwthread.c`: commit 95cb368095695fc233282264532704c7141d26ed ("Merge up to 1f3f635693a1ddc85f362dc324cb49c3e7b75f27 from upstream") and commit c545b9c4ab32c5566091dfe05e7f1765c39ed069 ("rtos/hwthread: use printf format specifier") -- resolved in favour of upstream version.
* `src/rtos/rtos.c`: commit 737f013ada404fdd620c2c0d4f1cb318a0ff64b2 ("Support RV32/RV64 mainline/metal stackings (#586)") and commit b3b790e4e01331f79b196f8312193ae0130d2394 ("rtos: rework rtos_create()") -- adjusted the handlers, adopted the error checks.
* `src/server/gdb_server.c`: commit 52c9ae0aa1f1413e01e3599fc23b9e98147129b6 ("server/gdb_server: Handle events if first target is unavailable") and commit 7f57e72afec76fcbc6210ddbd85ad4ecd5062ab5 ("gdb_server: Operate on available targets.") -- resolved in favor of the version from RISC-V OpenOCD, since it is being upstreamed.
Link: https://review.openocd.org/c/openocd/+/8914
* `src/target/breakpoints.c`: commit b1f3a7581928f66f8e87a1cb4307065068747088 ("target/riscv: Don't resume unavailable harts.") and commit ddef9cf73b1a39a2cf07ab04c8aee3aae1caae98 ("target: align switch and case statements") -- aligned the statements.
* `src/target/riscv/riscv.c`: due to commit 325e6d38b5d3edb3563481b96c897c3dc6f84bd8 ("target: Use 'bool' data type in mmu()") -- adjusted to use `bool` instead of `int`.
* `src/target/target.c`: commit 41b5b5471b010ad8b6ef24125b20eab2260278de ("Revert "break from long loops on shutdown request"") and commit 9fe3780432847b717d308a224adc4bd50716976e ("openocd: drop iteration downsampling for keep_alive()") -- also dropped the downsampling.
* `src/helper/base64.[ch]`: due to commit 6a3abda0b46a777490a8ce5bc748e6fed6775354 ("helper: add base64 encoding/decoding helpers from FreeBSD") -- resolved in favour of upstream.

aap-sc committed:
[syntacore] tcl/interface: proper integration of digilent-hs2a.cfg

Change-Id: I4cdea73584b214beb88436c7390915567cda21df

aap-sc committed:
[quick fix] remove deprecated registers

Change-Id: I213b77ad7a1a7036dd377968f769a57a50037d41

aap-sc committed:
[syntacore] Syntacore OpenOCD scripts

This commit includes a number of scripts and utilities that are specific
to Syntacore OpenOCD distribution and are unlikely to be included in any
upstream repo.

Signed-off-by: Parshintsev Anatoly <anatoly.parshintsev@syntacore.com>
Co-authored-by: Kirill Radkin <kirill.radkin@syntacore.com>
Co-authored-by: Evgeniy Naydanov <evgeniy.naydanov@syntacore.com>
Change-Id: Ia42adea528f2869df27b6ac22876dab40e641d2a

aap-sc committed:
[syntacore] do not use guess-rev.sh to determine the version

Change-Id: I42ea6424af326e3ef2a5a7dc590eee3557746ad2

kr-sc committed:
[to upstream] Implement ability to re-examine target

Link: https://github.com/riscv-collab/riscv-openocd/pull/1020

Add a new OpenOCD command `re_examine` to enforce re-examination of target

Change-Id: I3e33b5c15077e41dedf09258cc6f16a54d863aa8
Signed-off-by: Kirill Radkin <kirill.radkin@syntacore.com>

kr-sc committed:
[cherry-pick] server/gdb_server: Add more error messages in gdb_new_connection

Link: https://review.openocd.org/c/openocd/+/7927/2

Change-Id: I78cccfb297ddf7be418da62f2ed17e12af97b5d1
Signed-off-by: Kirill Radkin <kirill.radkin@syntacore.com>

aap-sc committed:
[to upstream] Revert "Revert "target/image: zero-initialize ELF segments up to p_memsz""

This reverts commit 144d940a8b0276d2ecb97ee6615d8a78faf4117d.

Change-Id: I960227e47c2bd459aacafa24dd65c9c4351f8485

aap-sc committed:
[quick fix] add warning if memory read/write fail on a non-halted target

The OS change is abandoned.
Link: https://review.openocd.org/c/openocd/+/8167

Change-Id: I9340de33ffeff695c174bef3c2313a0779428b15

aap-sc committed:
[quick fix] quick-and-dirty workaround for somewhat confusing error message for the case when watchpoint is not found at address

Change-Id: I28588e7059dd4885a26c756819e2e81b71e396a5

en-sc committed:
[to upstream] fix access to unitialize variables

Link: https://review.openocd.org/c/openocd/+/8181

Change-Id: If987efcbb6cff299ed1d8b5d3ace3c558a4e299d

zapb-0 committed:
[cherry-pick] Drop libjaylink Git submodule

Link: https://review.openocd.org/c/openocd/+/8428/1

Building libjaylink as Git submodule is announced to be deprecated since
OpenOCD v0.13.0 release, drop it.

Change-Id: I209eda9c51930d187c99ac333f0a21fa577a2d16
Signed-off-by: Marc Schink <dev@zapb.de>

zapb-0 committed:
[cherry-pick] Drop jimtcl Git submodule

Link: https://review.openocd.org/c/openocd/+/8429/1

Building jimtcl as Git submodule is announced to be deprecated since
OpenOCD v0.13.0 release, drop it.

Change-Id: I550879af3e90f50b067bb77a18c77410975e8573
Signed-off-by: Marc Schink <dev@zapb.de>

en-sc committed:
[cherry-pick] jtag: add interface for multi-TAP scans

Link: https://review.openocd.org/c/openocd/+/8504

The functionality was supported before
e018c7c1d29e8dabb9b4a90bb9eb3574eb1668bb,
although the interface was different.

This commit adds a separate interface for multi-TAP scans, leaving the
common use-case of a single active TAP intact.

Change-Id: I4fa6009a5bd49d8ae6170fe7239010f9259dd064
Signed-off-by: Evgeniy Naydanov <evgeniy.naydanov@syntacore.com>

en-sc committed:
[cherry-pick] jtag: add `jtag execute ...` TCL command

Link: https://review.openocd.org/c/openocd/+/8505

The command allows:
* To execute multiple scans in a single queue.
* Execute scans on multiple TAPs.

Change-Id: I722bbe8955ea05222ebca7f20e1324b1a8a39fff
Signed-off-by: Evgeniy Naydanov <evgeniy.naydanov@syntacore.com>

en-sc committed:
[cherry-pick update] jtag: add `jtag execute ...` TCL command

This is an update of a cherry-picked commit 20fd300659d0 ("[cherry-pick]
jtag: add `jtag execute ...` TCL command").

Link: https://review.openocd.org/c/openocd/+/8505/9

The command allows:
* To execute multiple scans in a single queue.
* Execute scans on multiple TAPs.

Change-Id: I722bbe8955ea05222ebca7f20e1324b1a8a39fff
Signed-off-by: Evgeniy Naydanov <evgeniy.naydanov@syntacore.com>

aap-sc committed:
[cherry-pick] target/target: fix confusing error messages during verify_image

Link: https://review.openocd.org/c/openocd/+/8458/2

When calculating target memory checksum OpenOCD tries to use
target-specific routine for crc calculation on the target. If this
approach fails it uses target-independent fallback.

This commit fixes two issues:

1. It allows to use target-independent fallback even if the target does
   not provide target-specific routine (I assume this is a bug).
2. If target-specific routine fails, we print a warning indicating that
  fallback is used. Otherwise we had quite a confusing messages:

```
verify_image 512k 0x80000000
No working memory available. Specify -work-area-phys to target.
not enough working area available(requested 1112)
verified 524288 bytes in 5.484234s (93.359 KiB/s)
```

After the patch these become:

```
verify_image 512k 0x80000000
[riscv.cpu0] No working memory available. Specify -work-area-phys ...
[riscv.cpu0] not enough working area available(requested 1112)
[riscv.cpu0] target-specific checksum_memory routine failed, atte ...
verified 524288 bytes in 6.473983s (79.086 KiB/s)
```

Note: messages are truncated to pass checkpatch.

Change-Id: I23a7549251ba21987212708ef35cdad4c5d5661a
Signed-off-by: Parshintsev Anatoly <anatoly.parshintsev@syntacore.com>

en-sc committed:
[cherry-pick update] testing: add tests targeting JTAG scan commands

This is an update of a cherry-picked commit c43b2a219614 ("[cherry-pick]
testing: add tests targeting JTAG scan commands").

Link: https://review.openocd.org/c/openocd/+/8507/15

The tests can be run via `make check` when JTAG `dummy` adapter is
enabled.

Change-Id: I6097460b7991d831d7b12b5037b331ba6addeeb4
Signed-off-by: Evgeniy Naydanov <evgeniy.naydanov@syntacore.com>

en-sc committed:
[cherry-pick] testing: add tests targeting JTAG scan commands

Link: https://review.openocd.org/c/openocd/+/8507

The tests can be run via `make check` when JTAG `dummy` adapter is
enabled.

Change-Id: I6097460b7991d831d7b12b5037b331ba6addeeb4
Signed-off-by: Evgeniy Naydanov <evgeniy.naydanov@syntacore.com>

aap-sc committed:
[syntacore] implemented basic embargo facilities

if we need to embargo the codebase, we'll need:

```
rm src/target/embargo/sc-ext/*
touch src/target/embargo/sc-ext/Makefile.am
```

Change-Id: I5f03a5a8f91dadd0004cea2414ab870358115fab

en-sc committed:
[cherry-pick] jtag/drivers: extend dummy adapter to check scanned values

Link: https://review.openocd.org/c/openocd/+/8508

Specifying TCL procedures `::dummy::get_tdo` and
`::dummy::state_transition` customizes the behavior of Dummy JTAG
adapter and enables testing of scanned values without the need for extra
HW.

Change-Id: I0d232224f603c6aafbc4eb6db0ba61502c75a421
Signed-off-by: Evgeniy Naydanov <evgeniy.naydanov@syntacore.com>

en-sc committed:
[syntacore] target/riscv: introduce `scr5` target type

SCR 5 implements custom MMU with ability to directly access TLB. To
support this a new target type is introduced.

Change-Id: I0d6ed6d09a4e2063bc300092676986c17bc5def9
Signed-off-by: Evgeniy Naydanov <evgeniy.naydanov@syntacore.com>

en-sc committed:
[syntacore] target/riscv: make SCR5 TLB lookup optional

Add `-scr5-tlb-lookup` configuration option to control TLB lookup on
SCR5 targets.

Change-Id: Ief0758fc9d4f7e8496cd1f20ef3e48af79b844d2
Signed-off-by: Evgeniy Naydanov <evgeniy.naydanov@syntacore.com>

en-sc committed:
[syntacore] change version specification mechanism

Change-Id: I53e7f1455f21bc9d807876bc44a035d1d0b45b72
Signed-off-by: Evgeniy Naydanov <evgeniy.naydanov@syntacore.com>

en-sc committed:
[cherry-pick] helper/log: mark `fmt` argument of `alloc_vprintf()` as non-null

Link: https://review.openocd.org/c/openocd/+/8764/1

Building on Ubuntu 22.04 with `-fsanitize=undefined` (GCC 12.3.0)
results in an error:
```
In file included from /usr/include/stdio.h:894,
                 from /home/en-sc/openocd/src/helper/system.h:23,
                 from /home/en-sc/openocd/src/helper/replacements.h:18,
                 from /home/en-sc/openocd/src/helper/log.c:20:
In function ‘vsnprintf’,
    inlined from ‘alloc_vprintf’ at /home/en-sc/openocd/src/helper/log.c:347:8:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:85:10: error: null format string [-Werror=format-truncation=]
   85 |   return __builtin___vsnprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   86 |                                     __glibc_objsize (__s), __fmt, __ap);
      |                                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
```

The error mentiones the call site `src/helper/log.c:347`. There
`vsnprintf()` is called passing `fmt` as format string.

Mark the format string as non-null in `alloc_vprintf()`.

Change-Id: I91011490715998ef5a931c19c3c9d74a1a304e5d
Signed-off-by: Evgeniy Naydanov <evgeniy.naydanov@syntacore.com>

en-sc committed:
[backport] rtos/rtos: handle OOM in `rtos_thread_packet()`

Return an error in case `calloc()` fails.

Change-Id: Ibb21a62991be83be8b219887953ccf27156f8af5
Signed-off-by: Evgeniy Naydanov <evgeniy.naydanov@syntacore.com>
Reviewed-on: https://review.openocd.org/c/openocd/+/8763
Tested-by: jenkins
Reviewed-by: zapb <dev@zapb.de>

en-sc committed:
[backport] rtos/linux: handle OOM in `linux_gdb_thread_packet()`

Return an error in case `calloc()` fails.

Change-Id: Id1b758a1edbae3d71d625d1992579b99720d77d6
Signed-off-by: Evgeniy Naydanov <evgeniy.naydanov@syntacore.com>
Reviewed-on: https://review.openocd.org/c/openocd/+/8762
Tested-by: jenkins
Reviewed-by: zapb <dev@zapb.de>

en-sc committed:
[cherry-pick] rtos/linux: fix name overwrite in `linux_thread_extra_info()`

Link: https://review.openocd.org/c/openocd/+/8761/5

commit 908ee4dc9641bd3df2eb00264575501867da539d ("build: remove clang
unused variable assignment warnings") introduced an error:
```
-            tmp_str_ptr += sprintf(tmp_str_ptr, "%s", name);
+            sprintf(tmp_str_ptr, "%s", name);
             sprintf(tmp_str_ptr, "%s", temp->name);
```
This results in `name` being overwritten by `temp->name`.
Fix this, adding OOM handling along the way.

Change-Id: Id41f73247c3f7e6194d7c92187ad3163a9ea6c89
Signed-off-by: Evgeniy Naydanov <evgeniy.naydanov@syntacore.com>

tom-van committed:
[backport] drivers/linuxspidev: fix compilation on Linux older than 3.15

Although the commit [1] which introduced SPI_IOC_WR_MODE32 is 10 years
old, some hardware may enforce old Linux because the vendor
didn't bother with system updates.

Change-Id: I76d0b38c8646b1be329418860916b9f01b90990d
Link: [1] https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/include/uapi/linux/spi/spidev.h?h=linux-3.15.y&id=dc64d39b54c1e9db97a6fb1ca52598c981728157
Signed-off-by: Tomas Vanek <vanekt@fbl.cz>
Reviewed-on: https://review.openocd.org/c/openocd/+/8759
Tested-by: jenkins
Reviewed-by: Antonio Borneo <borneo.antonio@gmail.com>

en-sc committed:
[backport] src/target/riscv: error out of the `examine()` in case `abits` is zero

This is a backport of commit 45231117b8686af3b896c9645a7b5b7b3b8e1cfc
("src/target/riscv: error out of the examine() in case abits is zero")
Link: https://github.com/riscv-collab/riscv-openocd/pull/1238/

The spec requires `abits` to be no less then 7 ( [3.1. Debug Module
Interface (DMI)]):
> The DMI uses between 7 and 32 address bits

Commit a450a7d4966a19c48b2543810f5b8370af6383e5 ("Fix data types around
batch.{c,h}") introduced a check that issues a warning if `abits` is
less then 7. The reason it's a warning and not an error -- on Spike
`abits` is 6.

However, if the examination is to proceed when `abits` is zero, an
assert in `get_dmi_scan_length()` will trigger when attempting to
communicate over DMI.
Link: https://github.com/JanMatCodasip/riscv-openocd/blob/a450a7d4966a19c48b2543810f5b8370af6383e5/src/target/riscv/batch.c#L25

It's impossible to communicate over DMI with `abits == 0`, so an error
is returned instead of a warning in this case.

Change-Id: Ice4622adccfc8304b37bc678253cc19d8ac1457e
Signed-off-by: Evgeniy Naydanov <evgeniy.naydanov@syntacore.com>

en-sc committed:
[syntacore] tools/syntacore: add N extension registers to IDE cloack

* `N` extension is removed from RISC-V priveleged spec from 1.11 (covers
  `sedeleg`, `sideleg`).

* While I didn't find the exact spec version where `uscratchcsw`,
  `uscratchcswl` were removed, it seems like there are also a part of
  the deprecated N extension (from the simularity to `uscratch`).

Change-Id: Ife9db04be0644083633a902e0b1513780c9eb0d2
Signed-off-by: Evgeniy Naydanov <evgeniy.naydanov@syntacore.com>

en-sc committed:
Revert "[quick fix] remove deprecated registers"

This reverts commit 3d7902303a5fd2e504adc13e60ee6bf9906c310c ("[quick
fix] remove deprecated registers").
The reasoning for revert is as follows:
1. This is a change to an auto-generated file.
2. To adjust existence of registers `riscv_reg_impl_gdb_regno_exist()`
   should be used instead.

Change-Id: Ib5f569a64f453f85294de4476e5826c1c8620017
Signed-off-by: Evgeniy Naydanov <evgeniy.naydanov@syntacore.com>

aap-sc committed:
[syntacore] support for platform extensions in syntacore_target.cfg

Add `platform_extension` that allows to specify a key to be used with
`::scr_info::expose_custom_string` function.

Change-Id: Ib0b4e5a5a2e2a6f1eb0ab950e4e60e800e27f474
Signed-off-by: Parshintsev Anatoly <anatoly.parshintsev@syntacore.com>

aap-sc committed:
[syntacore] support for platform extensions in `openocd_run.exp`

Add the `--platform` argument to configure the `platform` parameter of
`syntacore_target.cfg`.

Change-Id: I45195e09a1fce0d8763bd4d43f44f8b79a35b376
Signed-off-by: Parshintsev Anatoly <anatoly.parshintsev@syntacore.com>

aap-sc committed:
[to upstream] target/riscv: do not attempt to read memory if semihosting is not enabled

We need to discuss with upstream folks feasibility of reverting relevant
https://github.com/riscv-collab/riscv-openocd/pull/1200
(b7d9ab5d1e201cd1135ec587099bffa100bee9ca).
In my personal opionion that change was quite dangerous since it may
affect stability of debug session. I decided to disable this in
Syntacore fork for now.

Change-Id: I48a5060f56cc0d85e229804e60696ea2b8d5a5a8

en-sc committed:
[quick fix] Allign the backport from mainline

Commit 022e438292de992cc558f268b2679e76ad313db6 ("target: Change policy
of removing watchpoints/breakpoints.") was improperly backported.
Link: https://review.openocd.org/c/openocd/+/7738
This commit aligns the versions.

Change-Id: I595bf92c71a3d4e2ecf6380ec054434820c64c55

AlexandraKulyatskaya committed:
Add riscv-specific command to access RISC-V CSRs by direct specification address

Implementation of two RISC-V-specific commands for direct reading and
writing of CSRs.

Change-Id: I0775e4796e00b372789f404e53d2fa698ab2f78d

en-sc committed:
[syntacore] tcl/syntacore: do not modify `*counteren` in HPM setup

The RISC-V Instruction Set Manual: Volume II: Privileged Architecture
Version 20241101 [3.1.11. Machine Counter-Enable (mcounteren) Register]
states:
> The settings in this register only control accessibility. The act of
reading or writing this register does not affect the underlying
counters, which continue to increment even when not accessible.

Therefore, there is no need to adjust it's value when setting-up HPM
counters.

Change-Id: I083912ce3a8762ef33f805b71baa099cbf3d6a78
Signed-off-by: Evgeniy Naydanov <evgeniy.naydanov@syntacore.com>

aap-sc committed:
[ci] add expect package

Syntacore build system requires expect package to generate
`syntacore_ide_csr_cloak.cfg` file. Adjust pipeline to include the tool.

Change-Id: I9782a07d8b0e653e6800d005b8a187823ce0a013
Signed-off-by: Parshintsev Anatoly <anatoly.parshintsev@syntacore.com>

aap-sc committed:
Merge pull request #2 from syntacore/aap-sc/ci_expect

add package expect to ci

en-sc committed:
Merge pull request #1 from syntacore/en-sc/drop-xcounteren

[syntacore] tcl/syntacore: do not modify `*counteren` in HPM setup

en-sc committed:
Merge pull request #3 from syntacore/en-sc/revert-encoding

Revert "[quick fix] remove deprecated registers"

en-sc committed:
Merge pull request #4 from syntacore/en-sc/use-scr-info

Support Syntacore platforms

en-sc committed:
[syntacore] drop Syntacore extensions

Remove the `--enable-syntacore-extensions` configure argument and
related machinery.

Change-Id: I9c0251af6490b22bf26ef93c6e11cc30518142d4
Signed-off-by: Evgeniy Naydanov <evgeniy.naydanov@syntacore.com>

en-sc committed:
Merge pull request #6 from syntacore/en-sc/drop-sc-ext

[syntacore] drop Syntacore extensions

en-sc committed:
Merge up to 6f84e90d5cc0e98f895ea0fa31212326b50ab769 from RISC-V OpenOCD

* `configure.ac`: commit d24b3425b080ba3c488b82a46eb75a3077ffd5b3
  ("[cherry-pick] Drop libjaylink Git submodule") and commit
  9eb2426411aa0c1c13f5fc59801ac3cb3319f476 ("configure.ac: show the
  Remote Bitbang driver in the config summary").
* `src/Makefile.am`: commit 033718c20ae86e584c27fb9ce281172e22ccf756
  ("[syntacore] change version specification mechanism") and commit
  6834f022b96fb1c7f5829166578e01a0ac223cb0 ("Makefile.am: Use
  SOURCE_DATE_EPOCH environment variable"). Resolved by using the
  version from Syntacore OpenOCD.
* `src/helper/log.h`: commit 9b85208a501b5bebc397629cc7ee0454c7bbfa63
  ("[cherry-pick] helper/log: mark `fmt` argument of `alloc_vprintf()`
  as non-null") and commit e12ceddd5ee4d946107e3764d05ce2810befb293
  ("helper/log: mark `fmt` argument of `alloc_vprintf()` as format
  string"). Resolved by using the version from upstream.
* `src/rtos/linux.c`: commit 754ca1f124774b833a983186e08d9640d2b13885
  ("[cherry-pick] rtos/linux: fix name overwrite in
  `linux_thread_extra_info()`") and commit
  16c6497a89600ab8e8b354e2fc2c0ceb9ae74330 ("rtos/linux: fix name
  overwrite in `linux_thread_extra_info()`"). Resolved by using the
  version from upstream.
* `src/target/riscv/riscv.c`: commit
  ee1239f8c37ef8715feb153ac7175f15d92f195f ("[syntacore] target/riscv:
  introduce `scr5` target type") and commit
  297844cf46c7de5faa9de1c4a9f223b505dc3e9c ("target: Use 'bool' data
  type in target_{step,resume}"). Resolved by using the `bool` type.

Change-Id: Ie1f4c49072ad1b877500589fcd3d0fbdc9160bf7
Signed-off-by: Evgeniy Naydanov <evgeniy.naydanov@syntacore.com>

aap-sc committed:
fix SV57 translation for kernel address space

In SV57 **virtual** addresses look as follows:

+--------+---------+---------+---------+---------+---------+-----------+
|copy of |  VPN[4] |  VPN[3] |  VPN[2] |  VPN[1] |  VPN[0] |  OFFSET   |
| bit 56 |  9 bits |  9 bits |  9 bits |  9 bits |  9 bits |  12 bits  |
+--------+---------+---------+---------+---------+---------+-----------+
| 63  57 | 56   48 | 47   40 | 39   32 | 31   24 | 23   12 | 11      0 |

while the structure of **physical** address is:

+------------+----------+----------+---------+----------+-----------+
|   PPN[4]   |  PPN[3]  |  PPN[2]  | PPN[1]  |  PPN[0]  |  OFFSET   |
|   8 bits   |  9 bits  |  9 bits  | 9 bits  |  9 bits  |  12 bits  |
+------------+----------+----------+---------+----------+-----------+
| 55      48 | 47    39 | 38    30 | 29   21 | 20    12 | 11      0 |

So the size of effective VA is 57, while PA is 56 bits.

When our translation routine constructs the final physical address, it
preserves a portion of the original virtual address (to handle superpage
cases). To mask out the part of the original VA that must be replaced
with the PPN from the PTE, the **virt2phys_info_t::pa_ppn_mask** field is
used.

In our codebase pa_ppn_mask[4] was initialized incorrectly for SV57
mode, resulting  in a single bit from the original VA to leak through,
producing an incorrect physical address.

aap-sc committed:
Merge pull request #8 from syntacore/aap-sc/sv57_fixup

fix translation of kernel addresses in sv57 mode

MarekVCodasip reviewed (changes requested):
Seems that you have missed a change from `int` to `bool` here: https://github.com/riscv-collab/riscv-openocd/pull/1290/files#diff-b4aa16f9e42cb8f0934baa7c8e0ec9c70a369bef98b99b26ae2e896c8aa95d6aR3413 (line 3413 of `riscv.c`)

en-sc committed:
Merge pull request #7 from syntacore/en-sc/from_upstream

Merge up to 6f84e90d5cc0e98f895ea0fa31212326b50ab769 from  RISC-V OpenOCD

en-sc committed:
Merge pull request #10 from AlexandraKulyatskaya/riscv-specific-direct-access-commands

RISC-V specific direct access commands

en-sc committed:
Merge pull request #5 from syntacore/en-sc/backport-jtag-exec-9

Update `jtag execute`-related patches.

en-sc committed:
Revert "[syntacore] tools/syntacore: add N extension registers to IDE cloack"

This reverts commit 00f4386d536379fefe71c90803c8839f2c861336.

The list in `ide_cloak.tcl` is a list of registers that should be
exposed, so adding registers to that list to hide them is a mistake.

Change-Id: I7c76aedc99f86949acbf05300af5dfa8ec5c9697
Signed-off-by: Evgeniy Naydanov <evgeniy.naydanov@syntacore.com>

en-sc committed:
Merge pull request #11 from syntacore/en-sc/revert-n-ext-in-cloak

Revert "[syntacore] tools/syntacore: add N extension registers to IDE…

aap-sc committed:
[cherry-pick] fix error detection during trigger removal

The error manifests on designs that allow CSR reads when hart is not
halted. This results in `rwp` command to erroneously report "success"

See: https://github.com/riscv-collab/riscv-openocd/pull/1296

Signed-off-by: Parshintsev Anatoly <anatoly.parshintsev@syntacore.com>

aap-sc committed:
Merge pull request #12 from syntacore/aap-sc/rwp_fixup

[cherry-pick] fix error detection during trigger removal

aap-sc committed:
[cherry-pick] target/riscv: fix potential UB reported by ubsan in ac cache lookup

when running OpenOCD built with ubsan enabled I've encountered the
following error message:

```
riscv-013.c:204:9: runtime error: null pointer passed as argument 2, ...
```

Checkpatch-ignore: GIT_COMMIT_ID

This is caused by a NULL pointer passed to bsearch function when
abstract command cache is still empty. This behavor was introduced in
commit ab97974d1b0d ("target/riscv: implement abstract command cache ...")

See: https://github.com/riscv-collab/riscv-openocd/pull/1297

Signed-off-by: Parshintsev Anatoly <anatoly.parshintsev@syntacore.com>

aap-sc committed:
implemented facilities to inject ubsan/valgrind workarounds in the codebase

this commit adds a new define that allows to inject code with various
workarounds we don't want to see in prodcution but are still desired to
mask known issues with ubsan/valgrind.

Signed-off-by: Parshintsev Anatoly <anatoly.parshintsev@syntacore.com>

aap-sc committed:
Merge pull request #13 from syntacore/aap-sc/ub_fixups

implemented facilities to inject ubsan/valgrind workarounds in the codebase

aap-sc committed:
[cherry-pick] target/riscv: fix ub during instruction decode

A left shift operation caused an implicit integer promotion, triggering
the following UBSan error:

```
left shift of 254 by 24 places cannot be represented in type 'int'
```

NOTE: it seems that this code won't work correctly with BE targets,
however this is a general problem of the whole implementation anyway.

See: https://github.com/riscv-collab/riscv-openocd/pull/1299

Signed-off-by: Anatoly Parshintsev <anatoly.parshintsev@syntacore.com>

aap-sc committed:
sc_fpga_lib: improve error reporting of sc_lib_riscv_csr_impl

Minor corrections:
- report the value for the timeout when .busy bit is not cleared for a
  long time
- remove extra space from some error messages
- fixed typo in the name of sc_lib_riscv_encode_abstract_command
  function

aap-sc committed:
sc_fpga_lib: sc_fpga_riscv_csr_write/read should not be called when background poll is enabled

sc_fpga_riscv_csr_read and sc_fpga_riscv_csr_write should not be used in
contexts where background poll is enabled, since the current
implementation of poll can clobber data0/data1 registers.

aap-sc committed:
openocd_run.exp: match "." symbol at the end of rbb port

The RBB port should end with "." symbol, without it the script could
theoretically match the wrong (truncated) value.

aap-sc committed:
Merge pull request #17 from syntacore/aap-sc/openocd_run_rbb_port_fixup

openocd_run.exp: match "." symbol at the end of rbb port

aap-sc committed:
Merge pull request #16 from syntacore/aap-sc/sc_fpga_lib_csr_rw

sc_fpga_lib: minor corrections to sc_fpga_riscv_csr_read/write functions

aap-sc committed:
Merge pull request #14 from syntacore/aap-sc/ubsan_comparator_bsearch

[cherry-pick] target/riscv: fix potential UB reported by ubsan in ac cache lookup

aap-sc committed:
Merge pull request #18 from syntacore/aap-sc/garbage_watchpoint_ub

[cherry-pick] target/riscv: fix ub during instruction decode

aap-sc committed:
target/riscv: disable cache-on-read for VL and VSTART

Based On: https://review.openocd.org/c/openocd/+/9414

When VTYPE is updated:

- VL is set to some appropriate value
- VSTART is set to zero

Signed-off-by: Anatoly Parshintsev <anatoly.parshintsev@syntacore.com>

aap-sc committed:
target/riscv: vstart should be preserved when manipulating vector state

Based on: https://review.openocd.org/c/openocd/+/9415/2

read access should not ruin the current execution context

Signed-off-by: Anatoly Parshintsev <anatoly.parshintsev@syntacore.com>

aap-sc committed:
target/riscv: do not cache access to vtype

Based On: https://review.openocd.org/c/openocd/+/9416/2

Writes to the vtype register are expected to have the side effect of
resetting vl and vstart. However, the current implementation caches the
vtype state on read, which prevents subsequent writes with
already-cached values from being issued.

Signed-off-by: Anatoly Parshintsev <anatoly.parshintsev@syntacore.com>

aap-sc committed:
fix access to vector context

This fixes various issues raising from erroneous caching policy and incorrect restoration of vstart when accessing vector registers.

aap-sc committed:
target/riscv: do not cache access to vxrm, vxsat and vcsr

Based On: https://review.openocd.org/c/openocd/+/9416/2

vxrm and vxsat registers are mirrored in vcsr, so we can't cache any of
them.

Signed-off-by: Anatoly Parshintsev <anatoly.parshintsev@syntacore.com>

aap-sc committed:
Merge pull request #20 from syntacore/aap-sc/vxrm_vxsat_cache_fixup

target/riscv: do not cache access to vxrm, vxsat and vcsr

EccoTheDolphin committed:
target/riscv: always validate dmstatus.version field

Based On: https://review.openocd.org/c/openocd/+/9460

The dmstatus.version field indicates which revision of the RISC-V debug
specification the debug module implements. Since this field is always
readable, validating it on every dmstatus access provides a simple
and effective check that JTAG communication with the target is
functioning correctly.

Change-Id: I2bf40e92269e7ba3c747518e20f39069a1dee363
Signed-off-by: Anatoly Parshintsev <kupokupokupopo@gmail.com>

aap-sc committed:
Merge pull request #21 from syntacore/aap-sc/dmstatus_verification

target/riscv: always validate dmstatus.version field

eugoka committed:
Merge up to 6f84e90d5cc0e98f895ea0fa31212326b50ab769 from mainline OpenOCD

Change-Id: I95b2bb9690fa403df612ed83358f36f24535c297
